types/opt/value.go

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package opt
+
+import (
+	"fmt"
+	"reflect"
+
+	jsonv2 "github.com/go-json-experiment/json"
+	"github.com/go-json-experiment/json/jsontext"
+)
+
+// Value is an optional value to be JSON-encoded.
+// Unlike the specialized Bool type, an empty (unset) Value[T] is marshaled as
+// a JSON null regardless of the "omitempty" JSON tag option when using [encoding/json].
+// With [github.com/go-json-experiment/json], however, empty optional Values are
+// omitted from the output.
+type Value[T any] struct {
+	value T
+	set   bool
+}
+
+// Equal reports whether the receiver and the other value are equal.
+// If the template type T in Value[T] implements an Equal method, it will be used
+// instead of the == operator for comparing values.
+type equatable[T any] interface {
+	// Equal reports whether the receiver and the other values are equal.
+	Equal(other T) bool
+}
+
+// Empty returns an empty optional Value.
+func Empty[T any]() Value[T] {
```

Comment:
Personally, I find it more intuitive (and therefore more convenient) with symmetric approaches: if there's a function that returns a non-empty optional value, I would also expect a similar, symmetrically named function that returns an empty value. OTOH, this might not be idiomatic in Go, I like it somewhat less with `Of[T]` being renamed to `ValueOf[T]` (a symmetrical name would be `EmptyValue[T]` or alike, which is more characters), and a pair of `Value[T]{}`/`ValueOf[T](v T)` seem reasonable.

Reply:
I don't feel strongly about keeping this, but I'm not sure how often we'll use this.

The purpose of this type is to interact with Go structs and JSON serialization. In which case, one would have a Go struct on hand in a variable somewhere. I suspect that most times someone will do `f.OptionalField.Clear()` instead of `f.OptionalField = opts.Empty[FieldType]()`.

Up to you, though. You probably have a better vision of how this package will be used than I do.